src/Microsoft.AspNet.Razor/CodeGenerators/CSharpTagHelperCodeRenderer.cs

```diff
@@ -249,70 +205,135 @@ protected virtual string GenerateUniqueId()
-                        .WriteStringLiteral(attributeName)
-                        .WriteParameterSeparator()
-                        .Write(valueAccessor)
-                        .WriteEndMethodInvocation();
+                    // We need to capture the tag helper's property value accessor so we can retrieve it later
+                    // (if we need to).
+                    string valueAccessor = null;
+
+                    foreach (var associatedDescriptor in associatedDescriptors)
+                    {
+                        var associatedAttributeDescriptor = associatedDescriptor.Attributes.First(
+                            attributeDescriptor => attributeDescriptor.IsNameMatch(attribute.Key));
+                        var tagHelperVariableName = GetVariableName(associatedDescriptor);
+
+                        valueAccessor = RenderBoundAttribute(
+                            tagHelperVariableName,
+                            valueAccessor,
+                            attribute,
+                            associatedAttributeDescriptor);
+                    }
                 }
                 else
                 {
-                    // The attribute value has already been recorded, lets retrieve it from the stored value
-                    // accessors.
-                    _writer
-                        .WriteStartAssignment(valueAccessor)
-                        .Write(htmlAttributeValues[attributeName])
-                        .WriteLine(";");
+                    RenderUnboundAttribute(attribute);
+                }
+            }
+        }
+
+        private string RenderBoundAttribute(
+            string tagHelperVariableName,
+            string previousValueAccessor,
+            KeyValuePair<string, Chunk> attribute,
+            TagHelperAttributeDescriptor attributeDescriptor)
+        {
+            var attributeName = attribute.Key;
+            var attributeValueChunk = attribute.Value;
+            if (attributeValueChunk == null)
+            {
+                // Minimized attributes are not valid for bound attributes. TagHelperBlockRewriter has already
+                // logged an error if it was a bound attribute; so we can skip.
+                return null;
+            }
+
+            // We need to capture the tag helper's property value accessor so we can retrieve it later (if we need to).
+            var currentValueAccessor = string.Format(
+                CultureInfo.InvariantCulture,
+                "{0}.{1}",
+                tagHelperVariableName,
+                attributeDescriptor.PropertyName);
+
+            if (attributeDescriptor.IsIndexer)
+            {
+                var dictionaryKey = attributeName.Substring(attributeDescriptor.Name.Length);
+                currentValueAccessor += $"[\"{dictionaryKey}\"]";
+            }
+
+            // If there's no location to pull the attribute value from then we need to execute the attribute value.
```

Comment:
Alternatively use `// If this attribute value has not been seen before, need to record its value.`